daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversionStackSafe.scala

```diff
@@ -0,0 +1,368 @@
+import com.daml.lf.speedy.{SExpr0 => source}
+import com.daml.lf.speedy.{SExpr1 => target}
+
+import scala.annotation.tailrec
+
+private[speedy] object ClosureConversionStackSafe {
+
+  private[speedy] def closureConvert(source0: source.SExpr): target.SExpr = {
+
+    // This is the 'Env' implementation from the pre-stack-safe closure-conversion code.
+    // TODO: Recode the 'Env' management to avoid the polynomial-complexity of 'shift'.
+    case class Env(mapping: Map[Int, target.SELoc]) {
+
+      def lookup(i: Int): target.SELoc =
+        mapping.get(i) match {
+          case Some(loc) => loc
+          case None =>
+            throw sys.error(s"lookup($i),in:$mapping")
+        }
+
+      def shift(n: Int): Env = {
+        def shiftLoc(loc: target.SELoc, n: Int): target.SELoc = loc match {
+          case target.SELocS(i) => target.SELocS(i + n)
+          case target.SELocA(_) | target.SELocF(_) => loc
+        }
+        // We must update both the keys of the map (the relative-indexes from the original SEVar)
+        // And also any values in the map which are stack located (SELocS), which are also indexed relatively
+        val m1 = mapping.map { case (k, loc) => (n + k, shiftLoc(loc, n)) }
+        // And create mappings for the `n` new stack items
+        val m2 = (1 to n).map(i => (i, target.SELocS(i)))
+        Env(m1 ++ m2)
+      }
+    }
+
+    object Env {
+      def apply(): Env = {
+        Env(Map.empty)
+      }
+      def absBody(arity: Int, fvs: List[Int]): Env = {
+        val newRemapsF: Map[Int, target.SELoc] = fvs.zipWithIndex.map { case (orig, i) =>
+          (orig + arity) -> target.SELocF(i)
+        }.toMap
+        val newRemapsA = (1 to arity).map { case i =>
+          i -> target.SELocA(arity - i)
+        }
+        // The keys in newRemapsF and newRemapsA are disjoint
+        val m1 = newRemapsF ++ newRemapsA
+        Env(m1)
+      }
+    }
+
+    /** Closure-conversion, Traversal:
+      *
+      *   To ensure stack-safety, the input expression is traversed by a single tail-recursive 'loop'.
+      *   During the 'Traversal', we are either:
+      *   - going 'Down' a source expression (subtree), with an 'Env' for context, or
+      *   - coming 'Up' with a target expression (result for a subtree)
+      *
+      *   In both cases we have a continuation ('Cont') argument which says how to proceed.
+      */
+    sealed trait Traversal
+    object Traversal {
+      final case class Down(exp: source.SExpr, env: Env) extends Traversal
+      final case class Up(exp: target.SExpr) extends Traversal
+    }
+    import Traversal._
+
+    /** Closure Conversion, Cont:
+      *
+      *   The multiple forms for a continuation describe how the result of transforming a
+      *   sub-expression should be embedded in the continuing traversal. The continuation
+      *   forms correspond to the source expression forms: specifically, the location of
+      *   recursive expression instances.
```

Comment:
```suggestion
      *   recursive expression instances (values of type SExpr).
```
recusive expression instance sounded a bit confusing to me at first.

Reply:
done